### ISSUE TITLE
MAINT, CI: decide on formatting/linting for Python code

### ISSUE THREAD

**Issue by tylerjereddy** [darshan-hpc/darshan#409]
Fairly minor point, but we probably don't want debates/time wasted on whitespace/formatting issues, so eventually having i.e., CI do a linter check with whatever formatter people prefer may help avoid those arguments for the Python code. `pycodestyle`, `black`, several other popular options I think...

One other annoying decision is whether to apply it *en masse* to the Python code already present in the code base (pretty annoying to mass-reformat), or just to diffs that change in new PRs--the latter is a bit more work to make happen but less disruptive.

**Comment by jakobluettgau** [comment]
I also like `black` for not having to think much about how to format things and the results usually are quite sensible :)

Currently, the makefile comes with a target "make lint" using flake8 (which uses `pycodestyle`) for pydarshan with a few PEPs disabled, but it is not enforced by a pre-commit hook or anything like that. I wouldn't mind if we would add something like that though, considering we could always disable for a few selected lines that benefit from non-conforming formatting via `# fmt: off` and `# fmt: on`.

**Comment by carns** [comment]
I like the pre-commit method too.  We do something like this in several of the Mochi repos thanks to @mdorier :

https://github.com/mochi-hpc/mochi-margo/blob/main/maint/hooks/pre-commit

... with a top level prepare script that installs it (pre-commit hooks are local).

https://github.com/mochi-hpc/mochi-margo/blob/main/prepare.sh

I've grown accustomed to this method; it gives you a chance to clean up your code locally before pushing.  That example is using clang-format but of course it could just as easily be a python tool.  Would be best to start with just python code in Darshan at this point.

**Comment by nawtrey** [comment]
I think it would be nice to have pre-commit hooks for `black` and `flake8` setup in a way that we can run the same checks using the `Makefile`. We could have them run in the CI as well, but that might be redundant. 

[`pre-commit`](https://pre-commit.com/) has the capability of doing these things with relative ease. If we can get the linting configurations right, we can add an entry to the `Makefile` (maybe replace the `lint` entry) that executes `pre-commit run` and it will perform whatever hooks we add. It's a bit tricky to check only the staged files, so that will take some experimenting if we don't want to reformat every module.

**Comment by tylerjereddy** [comment]
I would copy the basic infrastructure i.e., SciPy uses for linting `git` diffs rather than re-formatting the whole code base and messing up `git` history, using their `lint_diff.py` module as a starting point (along with how they leverage it): https://github.com/scipy/scipy/blob/master/tools/lint_diff.py

I'm not sure I see a reason to use the `Makefile` infrastructure--many Python devs probably aren't even used to that. I'm not a huge fan of pre-commit hooks (I'd probably disable on my end anyway), but don't really care if that is added. The sanity check in CI should be fairly cheap to do so seems worth adding once a formatting approach is agreed upon.

**Comment by carns** [comment]
Ok.  The main capability that seems to be helpful is a way to check locally without pushing commits (doesn't necessarily have to be via git hook or makefile; I'll defer to you guys about whatever makes sense).

**Comment by jakobluettgau** [comment]
I think borrowing from the SciPy infrastructure sounds good.

I also agree, that reformatting the whole code base probably creates more headaches than slowly transitioning some files over time as they get touched.

As for the Makefile, maybe more generally then specifically for the case of linting/formatting, I see this mainly as a convenience entry point for common install or development actions. Also this is a relatively common practice actually with projects likes Pandas, PIL, PyTorch, scikit-learn, Seaborn...
If you say you'd encourage a contributor to run this against their stage and fix some worst offenders but be graceful for the most part, I think this is a good candidate to also include a target that runs some sensible default with the Makefile (it should really just take a few lines, maybe something like `git diff --name-only *.py | xargs -i sh -c "black {} || true"`)

Also consider that users of Darshan may be quite familiar or feel at home with things like `make test`, `make doc`, `make lint` coming from lower-level C/C++ in comparison to familiarizing with `pytest/nose/... <various args>`, `sphinx <various args>`, ... and countless other popular python tools :)

**Comment by tylerjereddy** [comment]
Fair enough, if we genuinely have C/C++ devs that want to work on the Python code regularly.

One other thing that is a bit annoying about the separation of the C and Python tooling is that the build happens using Makefile stuff/autotools and isn't orchestrated by Python itself--in other words `pip install .` should probably be able to build and link the `darshan-util` C code while it is building the Python sources. This is important for tools like `asv`, where the project should be able to detect the impact of changes in C code when rebuilding the Python project at different hashes, without having to manually hack commands to build the C code first (i.e., the Python build system becomes "first class" and orchestrates the full build chain).

At the moment, we have awkward separation where rebuilding the Python project from source is fairly strongly disconnected from the C shared library object--the source will just keep pointing to that same lib without rebuilding for changes in C.

**Comment by tylerjereddy** [comment]
(there is some stuff with `--with-extension` in `setup.py`, though I think (?) this is mostly for wheels vs. checking if C code has changed, etc.)
